lib/rubocop/cop/naming/memoized_instance_variable_name.rb

```diff
@@ -0,0 +1,47 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Naming
+      # This cop checks for memoized methods whose instance variable name
+      # does not match the method name.
+      #
+      # @example
+      #   # bad
+      #
+      #   # Method foo is memoized using an instance variable that is
+      #   # not `@foo`. This can cause confusion and bugs.
+      #   def foo
+      #     @something ||= calculate_expensive_thing
+      #   end
+      #
+      #   # good
+      #
+      #   def foo
+      #     @foo ||= calculate_expensive_thing
+      #   end
+      #
+      #   # good
+      #
+      #   def foo
+      #     @foo ||= begin
+      #       calculate_expensive_thing
+      #     end
+      #   end
+      #
+      class MemoizedInstanceVariableName < Cop
+        MSG = 'Memoized variable does not match method name.'.freeze
+
+        def on_def(node)
```

Comment:
I guess the cop should add an offesne to `defs` node. `defs` node is for `def self.foo`.